pkg/bpf/collection.go

```diff
@@ -41,6 +48,135 @@ func LoadCollectionSpec(path string) (*ebpf.CollectionSpec, error) {
 	return spec, nil
 }
 
+func RemoveUnreachableTailcalls(spec *ebpf.CollectionSpec) error {
+	return removeUnreachableTailcalls(spec)
+}
+
+func removeUnreachableTailcalls(spec *ebpf.CollectionSpec) error {
+	type TailCall struct {
+		referenced bool
+		visited    bool
+		spec       *ebpf.ProgramSpec
+	}
+
+	entrypoints := make([]*ebpf.ProgramSpec, 0)
+	tailcalls := make(map[uint32]*TailCall)
+
+	for _, prog := range spec.Programs {
+		var id, slot uint32
+		// Consider any program that doesn't follow the tailcall naming convention
+		// x/y to be an entrypoint.
+		if _, err := fmt.Sscanf(prog.SectionName, "%d/%v", &id, &slot); err != nil {
+			entrypoints = append(entrypoints, prog)
+			continue
+		}
+
+		const (
+			cilium_calls_map = 2
+		)
+
+		// Any program that does follow the x/y naming convention but not part
+		// of the cilium_calls map is also considered an entrypoint.
+		if id != cilium_calls_map {
+			entrypoints = append(entrypoints, prog)
+			continue
+		}
+
+		tailcalls[slot] = &TailCall{
+			spec: prog,
+		}
+	}
+
+	// Allow test_cilium_calls_[0-9]+, cilium_calls_[0-9]+, cilium_calls_hostns_[0-9]+, and cilium_calls_netdev_[0-9]+
+	// Compile in func and not in global scope to allow for the names to be updated.
+	callsMapRegex := regexp.MustCompile(fmt.Sprintf(
```

Comment:
Ummm, I'm not sure how to feel about this. First, panicking is bad. Second, stuffing the strings into a regex verbatim makes my spidey sense tingle.

I'm guessing we can't change the name of these maps without things breaking terribly?

Reply:
Yes, I hate this as well. The `test_` variant really skewed me while trying to get CI green. I currently don't really have any other way to identify static from dynamic maps other than name. And as you can see we are horribly inconsistent with our map naming.

I think we can perhaps use the legacy iproute ID, but that only lasts until we remove that in favor of sane logic.

I honestly don't know why the name for this map has to be so dynamic, perhaps we should just hardcode the name of the `CALLS_MAP` to something and then enforce in CI that it doesn't get touched. I think we can work out in the loader which programs do and don't share a calls maps (which seems to be the purpose of the different names at the moment)

Reply:
I replaced this regex with a string.Contains using the the `callsmap.MapName` gloabal variable. That should have the same effect and be slightly less fishy. In the long term the idea is to make the name of this map hardcoded or to move towards defining the tail calls statically in the map definition. 

For now this is the best we can do.